### DISCUSSION TITLE
using multiple frames for inference to fix identity swaps

### DISCUSSION THREAD

**Discussion by RubenTeunisse** [talmolab/sleap#1334] [Help!]
Hi!
When working with sleap I notice that the incorrect inferences are often for frames where I as a human annotator also have a hard time recognizing the body part positions. So that is great, it means it is nearing human-level classification. However, when I look at a couple frames before and after it suddenly becomes very clear where the body parts are. Is there an option in sleap to train a model that takes multiple frames as input for a single frame prediction (like a backbone with an lstm layer for example)? If not, is this a feature you are considering to implement? I think it would greatly improve performance. 

The reason I am asking is because we still get quite some identity swaps when applying tracking. This happens most often during these frames with close animal interaction where the model confuses body parts from one animal for those of another. I have already tried various configurations of the tracking parameters. Unfortunately, the bottom_up_id models are also not the solution because we cannot recognize the mice identities over multiple recordings. Do you have any tips or tricks for this?

Cheers,
Ruben

**Comment by talmo** [comment]
That's a great suggestion Ruben! This is a common problem and one that's pretty tricky to solve across all cases.

We're working on improving our tracking by using learned models, but we're still in development stage with that project.

I know of some other projects that use recurrence or temporal convolutions to add context from other frames, but it's not something we've experimented with ourselves yet.

This would be a great project for a student, so I'll see if we can find someone to give it a go. If we make any progress we'll update this thread.

If you have any other suggestions just let us know!

In the meantime, if you can't use ID models you don't have that many options unfortunately. Standard tricks to improve the poses during close interactions might help (improving imaging conditions, labeling more frames during those periods, tweaking the model) but it sounds like you've given these a try already. If the poses are as good as they'll get, then you might want to try and fix it in postprocessing using heuristics like the scores, distances or a statistical model to exclude and then interpolate over outliers.

If you give us a bit more info on what you've tried so far, we can help brainstorm more ideas.

Hope this helps!

Talmo

**Comment by RubenTeunisse** [comment]
Hi Talmo,

Thanks for the quick reply and great to hear you see a future for my suggestion. 

Because of your suggestion, I added over 200 frame annotations for close-interaction situations (total of 1500 labeled frames) and while it seems to have improved a little bit, it is not nearly enough. 

We have discussed improvements for imaging conditions, but these will only show in future datasets.

Fixing tracks in post processing is definitely an option I will further explore. 

A little more information about our setup and attempts:
We have about 4 hours of recordings of three black mice on a white background. For training I have tried various combinations of data augmentations from no to all available options. 
![image](https://github.com/talmolab/sleap/assets/32301392/bcd3c90e-7820-49b8-90d1-7e009bb2f9bd)
Receptive field looks good to me:
![image](https://github.com/talmolab/sleap/assets/32301392/5f066a09-6d4e-4b25-b846-b4a006cc618e)
Only  tried unet and resnet152, and have not touched the filters rate, middle block, up interpolate, sigma, loss weight, etc. 

Since the trained models show large deviations from the annotations for inference on difficult frames that occur in the training data (suggesting underfitting), I tried more filters as well as resnet152.

For tracking I have tried a lot of combinations, but settled with (following the options in GUI v3): flow tracking, cull to 3, th=0.8, instance, hungarian, 5, 0.95, save, no filters (kalman filters gave some error, do you think that is promising to look further into?), connect.

Please let me know if you have more suggestions or need particular info to help brainstorming.

Cheers,
Ruben

**Reply by Lauraschwarz** [reply]
heya, sorry if this is not relevant but i read your post and i got excited because i am using sleap in a similar setting to you :)
i noticed exactly what you were describing so i adjusted my labelling approach a little bit. i picked out the frames that my model is struggling with and i labelled 3 frames before and 3 frames after this target frame. for me this majorly improved the performance in those specific situations. (i guess it is quite obvious...) 

what also helped me was to run the tracking and apply the same method for the frames where new tracks spawn. this is usually when the predictions are so off, that the model struggles to keep track... 

my model is not perfect yet but its getting there :) (i only labelled 500 frames so far...)
